3rd_party/3rd_party.cmake

```diff
@@ -0,0 +1,267 @@
+    set(TORCH_LOCATION "/usr/local/gcc103/lib")
+    set(TORCH_EXTENSION ".so")
+  elseif("$ENV{CPP_CROSS_COMPILE}" STREQUAL "macosx")
+    message(STATUS "Copying macOS 3rd party libraries")
+    set(SYSROOT "/usr/local/sysroot-x86_64-apple-macosx10.14")
+    set(BOOST_LOCATION "${SYSROOT}/usr/local/lib")
+    set(BOOST_COMPILER "clang")
+    set(BOOST_EXTENSION "mt-x64-1_77.dylib")
+    set(BOOST_LIBRARIES "atomic" "chrono" "date_time" "filesystem" "iostreams" "log" "log_setup" "program_options" "regex" "system" "thread" "unit_test_framework")
+    set(XML_LOCATION)
+    set(GCC_RT_LOCATION)
+    set(STL_LOCATION)
+    set(OMP_LOCATION)
+    set(ZLIB_LOCATION)
+    set(TORCH_LIBRARIES "libtorch_cpu" "libc10")
+    set(TORCH_LOCATION "${SYSROOT}/usr/local/lib")
+    set(TORCH_EXTENSION ".dylib")
+  else()
+    message(STATUS "Copying Linux 3rd party libraries")
+    set(SYSROOT "/usr/local/sysroot-$ENV{CPP_CROSS_COMPILE}-linux-gnu")
+    set(BOOST_LOCATION "${SYSROOT}/usr/local/gcc103/lib")
+    set(BOOST_COMPILER "gcc")
+    if("$ENV{CPP_CROSS_COMPILE}" STREQUAL "aarch64")
+      set(BOOST_ARCH "a64")
+    else()
+      message(FATAL_ERROR "Cannot cross compile to $ENV{CPP_CROSS_COMPILE}")
+      return()
+    endif()
+    set(BOOST_EXTENSION "mt-${BOOST_ARCH}-1_77.so.1.77.0")
+    set(BOOST_LIBRARIES "atomic" "chrono" "date_time" "filesystem" "iostreams" "log" "log_setup" "program_options" "regex" "system" "thread" "unit_test_framework")
+    set(XML_LOCATION "${SYSROOT}/usr/local/gcc103/lib")
+    set(XML_EXTENSION ".so.2")
+    set(GCC_RT_LOCATION "${SYSROOT}/usr/local/gcc103/lib64")
+    set(GCC_RT_EXTENSION ".so.1")
+    set(STL_LOCATION "${SYSROOT}/usr/local/gcc103/lib64")
+    set(STL_PATTERN "libstdc++")
+    set(STL_EXTENSION ".so.6")
+    set(OMP_LOCATION "${SYSROOT}/usr/local/gcc103/lib64")
+    set(OMP_PATTERN "libgomp")
+    set(OMP_EXTENSION ".so.1")
+    set(ZLIB_LOCATION)
+    set(TORCH_LIBRARIES "libtorch_cpu" "libc10")
+    set(TORCH_LOCATION "${SYSROOT}/usr/local/gcc103/lib")
+    set(TORCH_EXTENSION ".so")
+  endif()
+else()
+  message(STATUS "Copying Windows 3rd party libraries")
+  set(LOCAL_DRIVE "C:")
+  if(DEFINED ENV{LOCAL_DRIVE})
+    set(LOCAL_DRIVE $ENV{LOCAL_DRIVE})
+  endif()
+  # These directories are correct for the way our Windows 2016 build
+  # server is currently set up
+  set(BOOST_LOCATION "${LOCAL_DRIVE}/usr/local/lib")
+  set(BOOST_COMPILER "vc")
+  set(BOOST_EXTENSION "mt-x64-1_77.dll")
+  set(BOOST_LIBRARIES "chrono" "date_time" "filesystem" "iostreams" "log" "log_setup" "program_options" "regex" "system" "thread" "unit_test_framework")
+  set(XML_LOCATION "${LOCAL_DRIVE}/usr/local/bin")
+  set(XML_EXTENSION ".dll")
+  set(GCC_RT_LOCATION)
+  # Read VCBASE from environment if defined, otherwise default to VS Professional 2019
+  set(VCBASE "Program Files (x86)/Microsoft Visual Studio/2019/Professional")
+  if (DEFINED ENV{VCBASE})
+    set(VCBASE $ENV{VCBASE})
+  endif()
+  set(VCVER "14.29.30133")

```

Comment:
I just noticed this.  Presumably it's correct for the current CI workers, but it changes on every Visual Studio patch so we cannot have this hardcoded long term.

The way we used to get it was `/bin/ls -1 /c/PROGRA~2/MICROS~2/2019/Professional/VC/Tools/MSVC | tail -1`.

@edsavage please can you think of a way to get this dynamically in the future.  Effectively it needs to list the sub-directories of `Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC` in sorted order and set `VCVER` to the name of the last one.

Reply:
We can use `CMake`'s `file(GLOB)` command and  then selecting the last item in the returned list. In fact we already do that in `cmake\compilers\vs2019.cmake`, e.g.
```
file(GLOB MSVC_VERS "${VCBASE}/VC/Tools/MSVC/*")
list(GET MSVC_VERS -1 MSVC_VER)
if(${MSVC_VER} MATCHES "/([^/]+)$")
  set(VCVER ${CMAKE_MATCH_1})
endif()
```